### PR TITLE
review: fix: Reject invalid float/double literals

### PR DESCRIPTION
**Pull request by I-Al-Istannen**
## About
`float` and `double` can represent ~~three~~ *many* values you can not write using a floating literal. Some examples are `Infinity`, `-Infinity` and then the whole family of `NaN` values.

As they can not be written down directly, the current printer creates invalid code for them. Forbidding such values in `CtLiteral` is not an option, as we can no longer represent `get(Double.class).getField("POSITIVE_INFINITY")`. 

Alternatively, we could also adjust the printer to *handle* (and not *reject*) such literals, by producing either reads of `Double.POSITIVE_INFINITY` or emitting `1d/0`.

## Further work and questions
I think we should probably offer a convenience method somewhere, which translates a `float` or `double` to a Literal or appropriate field read (`Double.POSITIVE_INFINITY`). Something to think about here is NaN-Boxing: Theoretically we need to use `Double.longBitsToDouble` or equivalent to preserve them.

### CODE REVIEW

File: src/main/java/spoon/support/reflect/code/CtLiteralImpl.java [hunk 1]
```diff
@@ -38,8 +38,8 @@ public T getValue() {
 
 	@Override
 	public <C extends CtLiteral<T>> C setValue(T value) {
```
[the remark this answers is not in the document]
Comment (slarse) [reply src/main/java/spoon/support/reflect/code/CtLiteralImpl.java:40]: YES, exactly, finally someone other than myself pointing this out!

[the remark this answers is not in the document]
Comment (slarse) [reply src/main/java/spoon/support/reflect/code/CtLiteralImpl.java:40]: @sonatype-lift ignore

File: src/main/java/spoon/reflect/visitor/LiteralHelper.java [hunk 2]
```diff
@@ -43,13 +44,19 @@ private static String getBasedString(Long value, LiteralBase base) {
 	}
 
 	private static String getBasedString(Float value, LiteralBase base) {
+		if (value.isInfinite() || value.isNaN()) {
```
Comment (slarse) [inline src/main/java/spoon/reflect/visitor/LiteralHelper.java:47]: How about a parameterized test for these values? Should be very little effort to see that we raise whenever the value is infinite or NaN. A test would also make the intent of the PR a little bit more clear.

File: src/main/java/spoon/reflect/visitor/LiteralHelper.java [hunk 3]
```diff
@@ -43,13 +44,19 @@ private static String getBasedString(Long value, LiteralBase base) {
...
+		}
 		if (base == LiteralBase.HEXADECIMAL) {
 			return Float.toHexString(value) + "F";
 		}
-		return Float.toString(value) + "F";
+		return value + "F";
 	}
 
 	private static String getBasedString(Double value, LiteralBase base) {
+		if (value.isInfinite() || value.isNaN()) {
```
Comment (slarse) [inline src/main/java/spoon/reflect/visitor/LiteralHelper.java:57]: Same as above.

**Review by slarse — CHANGES_REQUESTED** [review]
@I-Al-Istannen I think that forbidding this kind of output is the way to go, trying to print a value as a literal when that value cannot be represented as a literal is a fairly substantial logical error. The PR looks good as well in terms of production code, but tests would be nice to cement the intent :).

However, this PR is (ever so slightly) breaking. While I think it should be considered a bugfix and thus not warrant a major version bump, what do you think about flagging this as breaking? And on that note, our contributing guidelines have no set convention for marking a PR as breaking, so perhaps we should see to that.

**Comment by I-Al-Istannen** [comment]
> The PR looks good as well in terms of production code, but tests would be nice to cement the intent :).

Whoops, missed that. Thanks for pointing it out :)

> However, this PR is (ever so slightly) breaking. While I think it should be considered a bugfix and thus not warrant a major version bump, what do you think about flagging this as breaking?

Yea, that sounds fine. I'd have liked to remove the ability of CtLiteral to store such values at all, but that is a logic error. It is a bit unfortunate that `Double.POSITIVE_INFINITY`, for example, is represented by a literal that can not be printed. Such is the life of using reflection. Makes it a bit harder to track down where you screwed up though, if only the printing at the end fails… It took me a hot minute to find out why I had random `InfinityF` symbols in my code :D


---

What do you say about
> I think we should probably offer a convenience method somewhere, which translates a `float` or `double` to a Literal or appropriate field read (`Double.POSITIVE_INFINITY`).

My current code for that can be written in a nicer way but it still is really annoying how much you need to check:
```java
  @SuppressWarnings({"unchecked"})
  public static <T> CtExpression<T> getLiteral(Factory factory, Object value) {
    CtExpression<?> expression;
    if (value instanceof Float f && (f.isNaN() || f.isInfinite())) {
      expression = handleNanInfinite(factory, f);
    } else if (value instanceof Double d && (d.isNaN() || d.isInfinite())) {
      expression = handleNanInfinite(factory, d);
    } else {
      expression = factory.createLiteral(value);
    }
    return (CtExpression<T>) expression;
  }

  private static CtExpression<?> handleNanInfinite(Factory factory, Number number) {
    if (number instanceof Float f) {
      if (f.isNaN()) {
        return createFieldRead(factory, "NaN", factory.Type().FLOAT);
      }
      return f > 0
          ? createFieldRead(factory, "POSITIVE_INFINITY", factory.Type().FLOAT)
          : createFieldRead(factory, "NEGATIVE_INFINITY", factory.Type().FLOAT);
    } else if (number instanceof Double d) {
      if (d.isNaN()) {
        return createFieldRead(factory, "NaN", factory.Type().DOUBLE);
      }
      return d > 0
          ? createFieldRead(factory, "POSITIVE_INFINITY", factory.Type().DOUBLE)
          : createFieldRead(factory, "NEGATIVE_INFINITY", factory.Type().DOUBLE);
    } else {
      throw new IllegalArgumentException("Unexpected type: " + number.getClass() + " " + number);
    }
  }

  @SuppressWarnings({"unchecked", "rawtypes"})
  private static CtExpression<?> createFieldRead(
      Factory factory,
      String name,
      CtTypeReference<?> type
  ) {
    return factory.createFieldRead()
        .<CtFieldRead>setVariable(
            (CtFieldReference) type.getDeclaredField(name)
        )
        .setTarget(factory.createTypeAccess(type));
  }
```

**Review by slarse — APPROVED** [review]
@I-Al-Istannen Thanks, lgtm.

> I think we should probably offer a convenience method somewhere, which translates a float or double to a Literal or appropriate field read (Double.POSITIVE_INFINITY).

Perhaps. What would the return type of such a method be? `CtExpression<Float/Double>`? Would that be sufficient for the use case?

@monperrus FYI this PR is a bugfix but it is ever so slightly breaking (instead of producing invalid code, Spoon now throws an exception). We need to establish some practice for marking breaking pull requests.

**Comment by monperrus** [comment]
Ack.

> We need to establish some practice for marking breaking pull requests.

We have a label `breaking_pr`: https://github.com/INRIA/spoon/pulls?q=label%3Abreaking_pr+sort%3Aupdated-desc

**Comment by I-Al-Istannen** [comment]
> Perhaps. What would the return type of such a method be? CtExpression<Float/Double>? Would that be sufficient for the use case?

Yes, that would be fine. I am generating java code matching a runtime `float`/`double` literal.

**Comment by slarse** [comment]
@monperrus Right, I didn't actually mean marking the pull request but rather the commit itself after merge :). I.e. like we have `fix:`, we ought to also mark breaking commits with something.

@I-Al-Istannen Any suggestion for where such a helper method might live? Also, what exactly is the use case? It could potentially be argued that it's a bit of a fringe case to have floats and doubles that sometimes veer off into NaN or infinite territory when it comes to generating source code.

**Comment by I-Al-Istannen** [comment]
> Also, what exactly is the use case? It could potentially be argued that it's a bit of a fringe case to have floats and doubles that sometimes veer off into NaN or infinite territory when it comes to generating source code.

Maybe, I am fine with only having the code above in my project and not in spoon. The use case is performing computations / reading values from an object at runtime and using the result in generated code. The generated code then creates the same value. This is pretty similar to the shadow model.

It mostly felt like a logical extension of `createLiteral(object)`. It is a bit weird that you can put in an object and it dies at codegen. Throwing earlier is impossible, of course, as the shadow model *creates CtLiteral nodes with `NaN/INFINITY`* values.  
Maybe that's also something we should change and the shadow model should use such a method. Currently `factory.get(Double.class).getField("POSITIVE_INFINITY)` returns a CtLiteral with an unrepresentable literal. That sounds nice for partial evaluation/symbolic execution, but not as great for about anything else.

**Comment by slarse** [comment]
It may be worthwhile to have such helper methods simply to clarify that this is a problem that exists (it had never crossed my mind before). It's not like maintaining them will be a lot of effort.

Would you like to just submit your code in a PR and put them somewhere appropriate and we'll discuss it there? If we feel unsure about whether it should go in or not we can just call in the troops for a tiebreaker.

**Pull request merged**